### PR TITLE
Clarify the semantics of the uplink message timestamps

### LINKED ISSUE #5585: Clarify the semantics of the uplink message timestamps
### Summary

The timestamps found in the `UplinkMessage` message should have their semantics clarified.

### Current Situation

We currently have multiple timestamps available:

1. The timestamp at which the Gateway Server / Network Server received the frame:
https://github.com/TheThingsNetwork/lorawan-stack/blob/70d88dd86a9b17bf8faf1bbea6961d6c5d3f2ae7/api/messages.proto#L83-L85

2. The timestamp at which a singular transmission has been received, from the perspective of the gateway:
https://github.com/TheThingsNetwork/lorawan-stack/blob/70d88dd86a9b17bf8faf1bbea6961d6c5d3f2ae7/api/lorawan.proto#L326-L328

3. A set of timestamps of each individual transmission metadata:
https://github.com/TheThingsNetwork/lorawan-stack/blob/70d88dd86a9b17bf8faf1bbea6961d6c5d3f2ae7/api/metadata.proto#L41
https://github.com/TheThingsNetwork/lorawan-stack/blob/70d88dd86a9b17bf8faf1bbea6961d6c5d3f2ae7/api/messages.proto#L82

### Why do we need this? Who uses it, and when?

The semantics of timestamps 2 and 3 are not clear unfortunately. Specifically it is unclear what is the clock source of this timestamp (is it a GPS timestamp that the gateway is _sure_ of, or did it do some NTP and has only a bit of accuracy ?).

This has the following effects:
- Some uplinks are marked as coming from the 80s, because the gateway did some kind of NTP failure, or lacks a GPS fix.
- Some uplinks are buffered and the timestamp is mis-attributed (time of buffer submission vs actual receive timestamp).
- Correctly supporting `DeviceTimeAns` is not really possible (if we have 3 timestamps, which one do we pick ? per spec, it should be a GPS one)

### Proposed Implementation

_The proposals below are targeting the `RxMetadata` message_

Option 1: Change the semantics of the `time`, for uplink messages, such that the presence of the value signals that it is provided by GPS, and leave it empty otherwise.

PROS:
- It was always nullable, so taking it out is by the book correct.
- We guarantee that applications that use it have a useful timestamp.

CONS:
- De facto, this is a breaking change. Gateways that had 0 issues will suddenly no longer have a timestamp.

I don't think that we should go this way, but I'm documenting this here.

Option 2: Add a new field, `gps_time`, that is filled when the timestamp is really sourced by GPS. If we have both an NTP sourced timestamp, and a GPS sourced timestamp, `time` contains the GPS one.

PROS:
- Not a breaking change.
- It will clarify the semantics.

CONS:
- We add a new field to store. Keep in mind that this is both something we will carry around (data transfer between components) and store (event storage, NS recent uplinks storage maybe, AS storage integration), so we should be mindful.

I'm more in favor of this variant, but it's not entirely 'free' (storage / transfer costs). 

cc @johanstokking

### Contributing

- [X] I can help by doing more research.
- [X] I can help by implementing the feature after the proposal above is approved.
- [X] I can help by testing the feature before it's released.

### Code of Conduct

- [X] I agree to follow TTN's [Community Code of Conduct](https://github.com/TheThingsNetwork/Manifesto/blob/master/CodeOfConduct.md).

### PR DESCRIPTION
**Pull request by happyRip**
#### Summary

Closes #5585

#### Changes

- Added `gps_time` and `received_at` fields to `RxMetadata`
- Improved uplink message's `GatewayReceiveTime` accuracy (when possible)

#### Testing

Adjust relevant test cases to feature expected `received_at` and `gps_time` values.

##### Regressions

Possible regressions listed in #5585 and #5586.

#### Notes for Reviewers

nil

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/metadata.proto [hunk 1]
```diff
@@ -71,12 +71,16 @@ message RxMetadata {
   uint32 hopping_width = 19;
   // Frequency drift in Hz between start and end of an LR-FHSS packet (signed).
   int32 frequency_drift = 20;
+  // GPS sourced timestamp at the end of the transmission.
+  google.protobuf.Timestamp gps_time = 21;
+  // Time when the service received the message.
+  google.protobuf.Timestamp received_at = 22;
```
Comment (adriansmares) [inline api/metadata.proto:80]: I would add slightly more details to the comments here. We now are going to have 3 timestamps and that will be confusing in the absence of the explanation.

I would add a comment to `time` saying `Timestamp at the end of the transmission, provided by the gateway. The accuracy is undefined.`

For `gps_time` I would mention `Timestamp at the end of the transmission, provided by the gateway. Guaranteed to be based on a GPS PPS signal, with an accuracy of 1 millisecond.`.

For `received_at` I would mention `Timestamp at which the Gateway Server has received the message.`

File: pkg/gatewayserver/gatewayserver_test.go [hunk 2]
```diff
@@ -1365,6 +1365,7 @@ func TestGatewayServer(t *testing.T) {
 										for i, md := range msg.RxMetadata {
 											a.So(md.UplinkToken, should.NotBeEmpty)
 											md.UplinkToken = nil
+											expected.RxMetadata[i].ReceivedAt = md.ReceivedAt
```
Comment (adriansmares) [inline pkg/gatewayserver/gatewayserver_test.go]: I would say that it is better to erase it from `md` (set to `nil`).

File: pkg/gatewayserver/io/io.go [hunk 3]
```diff
@@ -297,6 +297,12 @@ func (c *Connection) HandleUp(up *ttnpb.UplinkMessage, frontendSync *FrontendClo
 		} else if !c.gateway.LocationPublic {
 			md.Location = nil
 		}
+
+		r := *ttnpb.StdTime(up.ReceivedAt)
+		if _, _, median, _, count := c.RTTStats(100, time.Now()); count > 0 {
+			r.Add(-median / 2)
```
Comment (adriansmares) [inline pkg/gatewayserver/io/io.go]: `time.Time.Add` does not modify `r` - you need to reassign`r`.

File: pkg/gatewayserver/io/io.go [hunk 4]
```diff
@@ -297,6 +297,12 @@ func (c *Connection) HandleUp(up *ttnpb.UplinkMessage, frontendSync *FrontendClo
 		} else if !c.gateway.LocationPublic {
 			md.Location = nil
 		}
+
+		r := *ttnpb.StdTime(up.ReceivedAt)
```
Comment (adriansmares) [inline pkg/gatewayserver/io/io.go]: It would be better to compute this outside of the loop, since `up.ReceivedAt` or the return of `RTTStats` does not change between iterations (mostly).

Also be careful do this write at the start of the loop - you have some `continue` statements above that will make the metadata not contain the `received_at` when triggered.

File: pkg/packetbrokeragent/translation.go [hunk 5]
```diff
@@ -80,15 +80,15 @@ func fromPBDataRate(dataRate *packetbroker.DataRate) (dr *ttnpb.DataRate, coding
...
-	// 	return &ttnpb.DataRate{
-	// 		Modulation: &ttnpb.DataRate_Lrfhss{
-	// 			Lrfhss: &ttnpb.LRFHSSDataRate{
-	// 				ModulationType:        mod.Lrfhss.ModulationType,
-	// 				OperatingChannelWidth: mod.Lrfhss.OperatingChannelWidth,
-	//  			CodingRate:            mod.Lrfhss.CodingRate,
-	// 			},
-	// 		},
-	// 	}, mod.Lrfhss.CodingRate, true
+	//	return &ttnpb.DataRate{
```
Comment (adriansmares) [inline pkg/packetbrokeragent/translation.go]: Revert the changes here.

File: pkg/packetbrokeragent/translation.go [hunk 6]
```diff
@@ -121,15 +121,15 @@ func toPBDataRate(dataRate *ttnpb.DataRate, codingRate string) (*packetbroker.Da
...
-	// 	return &packetbroker.DataRate{
-	// 		Modulation: &packetbroker.DataRate_Lrfhss{
-	// 			Lrfhss: &packetbroker.LRFHSSDataRate{
-	// 				ModulationType:        mod.Lrfhss.ModulationType,
-	// 				OperatingChannelWidth: mod.Lrfhss.OperatingChannelWidth,
-	// 				CodingRate:            mod.Lrfhss.CodingRate,
-	// 			},
-	// 		},
-	// 	}, true
+	//	return &packetbroker.DataRate{
```
Comment (adriansmares) [inline pkg/packetbrokeragent/translation.go]: Revert the changes here.

File: pkg/packetbrokeragent/translation.go [hunk 7]
```diff
@@ -373,11 +373,17 @@ func toPBUplink(ctx context.Context, msg *ttnpb.GatewayUplinkMessage, config For
...
 			}
 
-			if md.Time != nil {
-				t := ttnpb.StdTime(md.Time)
-				if gatewayReceiveTime == nil || t.Before(*gatewayReceiveTime) {
-					gatewayReceiveTime = t
-				}
+			var t *time.Time
+			switch {
+			case md.GpsTime != nil:
```
Comment (adriansmares) [inline pkg/packetbrokeragent/translation.go:383]: So this is what I requested in the issue, but maybe we need to do better - I think that if we have a GPS timestamp, we should override whatever we stored in `gatewayReceiveTime`, and no future updates to `gatewayReceiveTime` should be done.

Consider the following example:
2 metadata - one which contains a `gps_time` of `14:02:01.123123123` and one which contains a `time` of `14:02:01.0000000`. Since the secondary metadata is earlier we should use that, even though the GPS time is more precise and correct.

File: pkg/gatewayserver/io/io.go (hunk 4 above)
Comment (happyRip) [reply pkg/gatewayserver/io/io.go]: Turns out `receivedAt := *ttnpb.StdTime(up.ReceivedAt)` was declared outside of the loop, thus I [used it](https://github.com/TheThingsNetwork/lorawan-stack/pull/5598/commits/14ec453bfdafcd503df611ea71d9d9a2955aac89#diff-9c87734043add0ade22bca71f2a6641c1d37830e1aae533fbc6066899c479d1bR277).

File: pkg/packetbrokeragent/translation.go (hunk 7 above)
Comment (happyRip) [reply pkg/packetbrokeragent/translation.go:383]: Had multiple ideas on how to handle this case, [finally went with this](https://github.com/TheThingsNetwork/lorawan-stack/pull/5598/commits/14ec453bfdafcd503df611ea71d9d9a2955aac89#diff-a8a8e5ff5879a789dca4cdcd50830e79b514bf8eb5bd0ad553450bf7dcbb2a82R384).

```go
var t *time.Time
switch {
case md.GpsTime != nil:
  t = ttnpb.StdTime(md.GpsTime)
case md.Time != nil:
  t = ttnpb.StdTime(md.Time)
default:
  t = ttnpb.StdTime(md.ReceivedAt)
}
gatewayReceiveTime = t
```
`gatewayReceiveTime` at the very end emphasises the fact that's the variable I'd like to set here and it's going to be reassigned under every condition.

Other possible solutions:
- ```go
  switch {
  case md.GpsTime != nil:
    gatewayReceiveTime = ttnpb.StdTime(md.GpsTime)
  case md.Time != nil:
    gatewayReceiveTime = ttnpb.StdTime(md.Time)
  default:
    gatewayReceiveTime = ttnpb.StdTime(md.ReceivedAt)
  }
  ```
  - seems less clean due to extensive usage of a long variable
- ```go
  gatewayReceiveTime = func() *time.Time {
    switch {
    case md.GpsTime != nil:
      return ttnpb.StdTime(md.GpsTime)
    case md.Time != nil:
      return ttnpb.StdTime(md.Time)
    default:
      return ttnpb.StdTime(md.ReceivedAt)
    }
  }()
  ```
  - non-standard

Comment (adriansmares) [reply pkg/packetbrokeragent/translation.go:383]: But this does not solve the issue that I've mentioned - if the order of the metadata elements is 'low precision', 'high precision', 'low precision' you will keep the last timestamp, which is the low precision one.

Comment (happyRip) [reply pkg/packetbrokeragent/translation.go:383]: I see, so first choose most precise timestamp(s) and then choose among those in `gps` → `time` → `received_at` order?
Or should the earliest be chosen each case?

Comment (adriansmares) [reply pkg/packetbrokeragent/translation.go:383]: The axiom (which should be true in practice, but you never know) is that all `gps_time` is equivalent modulo 1 millisecond. So if you find a GPS time, then you don't need anything else. But if all of the metadata is missing the GPS time, then indeed we should take the earliest `time` and fallback to `received_at` (in case `time` is `nil`, compare using `received_at`).

Comment (happyRip) [reply pkg/packetbrokeragent/translation.go:383]: Solution proposal:

```go
gatewayReceiveTime = func() *time.Time {
	if t := md.GpsTime; t != nil {
		return ttnpb.StdTime(t)
	}

	t, r := ttnpb.StdTime(md.Time), ttnpb.StdTime(md.ReceivedAt)
	if t.IsZero() || r.Before(*t) {
		return r
	}
	return t
}()
```

Always use `gps_time` if available
```go
if t := md.GpsTime; t != nil {
  return ttnpb.StdTime(t)
}
```

Otherwise choose earliest from `time` and `received_at`.
Fallback to `received_at` if `time` is not available.
```go
t, r := ttnpb.StdTime(md.Time), ttnpb.StdTime(md.ReceivedAt)
if t.IsZero() || r.Before(*t) {
  return r
}
return t
```

Comment (adriansmares) [reply pkg/packetbrokeragent/translation.go:383]: The problem is that when you do this, you do not take into consideration that `gatewayReceiveTime` could already be set to a `gps_time`.

Take the following:
Metadata 1 contains `gps_time`, metadata 2 contains `time`.
When you process metadata 1 you will just set `gatewayReceiveTime` to `gps_time`.
Then you process metadata 2, there is no `gps_time`, so you go into the secondary condition and overwrite the timestamp.

What I have in mind is:
```
switch {
case hasGPSTime:
case md.GpsTime != nil:
  gatewayReceiveTime = ttnpb.StdTime(md.GpsTime)
  hasGPSTime = true
case gatewayReceiveTime == nil || md.Time != nil && minimal condition between gatewayReceiveTime and md.Time:
  gatewayReceiveTime = ttnpb.StdTime(md.Time)
case gatewayReceiveTime == nill || minimal condition between gatewayReceiveTime and md.ReceivedAt:
  gatewayReceiveTime = ttnpb.StdTime(md.ReceivedAt)
}
```

Comment (happyRip) [reply pkg/packetbrokeragent/translation.go:383]: Pushed a new solution:

Add `var hasGpsTime bool` before the loop begins

Inside the loop:
```go
timestampCase := func(t *pbtypes.Timestamp) bool {
	g := gatewayReceiveTime
	return g == nil || t != nil && ttnpb.StdTime(t).Before(*g)
}
switch {
case hasGpsTime:
case md.GpsTime != nil:
	gatewayReceiveTime = ttnpb.StdTime(md.GpsTime)
	hasGpsTime = true
case timestampCase(md.Time):
	gatewayReceiveTime = ttnpb.StdTime(md.Time)
case timestampCase(md.ReceivedAt):
	gatewayReceiveTime = ttnpb.StdTime(md.ReceivedAt)
}
```

File: pkg/gatewayserver/io/io.go [hunk 8]
```diff
@@ -274,6 +274,12 @@ func (c *Connection) HandleUp(up *ttnpb.UplinkMessage, frontendSync *FrontendClo
 	}
 
 	for _, md := range up.RxMetadata {
+		r := receivedAt
+		if _, _, median, _, count := c.RTTStats(100, time.Now()); count > 0 {
```
Comment (johanstokking) [inline pkg/gatewayserver/io/io.go]: Two things:
1. We want the message's received time as reference time. The difference between `time.Now()` and `receivedAt` is probably negligible, but still, it's cleaner to use `receivedAt` here I think
2. The result of the `c.RTTStats()` does not depend on the loop variable `md`, so this calculation should be moved to before the loop, and the loop is only needed if there's a known RTT to apply

File: pkg/packetbrokeragent/translation.go [hunk 9]
```diff
@@ -373,11 +374,19 @@ func toPBUplink(ctx context.Context, msg *ttnpb.GatewayUplinkMessage, config For
 				localization.Antennas = append(localization.Antennas, locAnt)
 			}
 
-			if md.Time != nil {
-				t := ttnpb.StdTime(md.Time)
-				if gatewayReceiveTime == nil || t.Before(*gatewayReceiveTime) {
-					gatewayReceiveTime = t
-				}
+			timestampCase := func(t *pbtypes.Timestamp) bool {
```
Comment (johanstokking) [inline pkg/packetbrokeragent/translation.go]:
```suggestion
			earlierGatewayReceiveTime := func(t *pbtypes.Timestamp) bool {
```

File: pkg/gatewayserver/io/io.go [hunk 10]
```diff
@@ -273,12 +273,13 @@ func (c *Connection) HandleUp(up *ttnpb.UplinkMessage, frontendSync *FrontendClo
 		panic("unreachable")
 	}
 
+	r := receivedAt
+	if _, _, median, _, count := c.RTTStats(100, receivedAt); count > 0 {
+		r = receivedAt.Add(-median / 2)
+	}
+	mdReceivedAtProto := ttnpb.ProtoTimePtr(r)
```
Comment (johanstokking) [inline pkg/gatewayserver/io/io.go]:
```suggestion
	receivedAtGateway := receivedAt
	if _, _, median, _, count := c.RTTStats(100, receivedAt); count > 0 {
		receivedAtGateway = receivedAt.Add(-median / 2)
	}
```

File: pkg/gatewayserver/io/io.go [hunk 11]
```diff
@@ -273,12 +273,13 @@ func (c *Connection) HandleUp(up *ttnpb.UplinkMessage, frontendSync *FrontendClo
...
 	for _, md := range up.RxMetadata {
-		r := receivedAt
-		if _, _, median, _, count := c.RTTStats(100, time.Now()); count > 0 {
-			r = r.Add(-median / 2)
-		}
-		md.ReceivedAt = ttnpb.ProtoTimePtr(r)
+		md.ReceivedAt = mdReceivedAtProto
```
Comment (johanstokking) [inline pkg/gatewayserver/io/io.go]:
```suggestion
		md.ReceivedAt = ttnpb.ProtoTimePtr(receivedAtGateway)
```

File: pkg/gatewayserver/io/io.go (hunk 10 above)
Comment (johanstokking) [reply pkg/gatewayserver/io/io.go]: Better not declare single character variables in a large scope to avoid confusion. In small scopes, like short loops etc, it's obviously permitted.

**Review by johanstokking — APPROVED** [review]

**Review by adriansmares — APPROVED** [review]
LGTM. Squash the last commit that was commited from the web UI.
